# Exercises
#### Do your work for this exercise in a python script or a jupyter notebook with the name dataframes.py or dataframes.ipynb.

##### For several of the following exercises, you'll need to load several datasets using the pydataset library. (If you get an error when trying to run the import below, use pip to install the pydataset package.)

In [1]:
from pydataset import data

When the instructions say to load a dataset, you can pass the name of the dataset as a string to the data function to load the dataset. You can also view the documentation for the data set by passing the show_doc keyword argument.

In [2]:
# data('mpg', show_doc=True) # view the documentation for the dataset
mpg = data('mpg') # load the dataset and store it in a variable

##### All the datasets loaded from the pydataset library will be pandas dataframes.

1. Copy the code from the lesson to create a dataframe full of student grades.

In [3]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)

pandas.core.frame.DataFrame

In [4]:
df.head()

name  math  english  reading
0  Sally    62       85       80
1   Jane    88       79       67
2  Suzie    94       74       95
3  Billy    98       96       88
4    Ada    77       92       98

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   name     12 non-null     object
 1   math     12 non-null     int64 
 2   english  12 non-null     int64 
 3   reading  12 non-null     int64 
dtypes: int64(3), object(1)
memory usage: 512.0+ bytes


In [6]:
df.describe()

math    english    reading
count  12.000000  12.000000  12.000000
mean   84.833333  77.666667  86.500000
std    11.134168  13.371158   9.643651
min    62.000000  62.000000  67.000000
25%    78.500000  63.750000  80.750000
50%    90.000000  77.500000  89.000000
75%    92.250000  86.750000  93.250000
max    98.000000  99.000000  98.000000

 * a. Create a column named passing_english that indicates whether each student has a passing grade in english.

In [7]:
df['english']

0     85
1     79
2     74
3     96
4     92
5     76
6     64
7     63
8     62
9     80
10    99
11    62
Name: english, dtype: int64

In [20]:
[df['english'] > 69]

[0      True
 1      True
 2      True
 3      True
 4      True
 5      True
 6     False
 7     False
 8     False
 9      True
 10     True
 11    False
 Name: english, dtype: bool]

In [14]:
df['english']['Pass' if df['english'] > 69 else 'Fail']

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [22]:
df['passing_english'] = df['english'] > 69

In [18]:
df['passing_english'] = lambda n: 'pass' if df['enlgish'] > 69 else 'fail'

In [23]:
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

 * b. Sort the english grades by the passing_english column. How are duplicates handled?

In [24]:
df[['name', 'english', 'passing_english']].sort_values('passing_english')

name  english  passing_english
6    Thomas       64            False
7     Marie       63            False
8    Albert       62            False
11     Alan       62            False
0     Sally       85             True
1      Jane       79             True
2     Suzie       74             True
3     Billy       96             True
4       Ada       92             True
5      John       76             True
9   Richard       80             True
10    Isaac       99             True

* sorted by the index

 * c. Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)

In [28]:
df[['name', 'english', 'passing_english']].sort_values('name').sort_values('passing_english')

name  english  passing_english
11     Alan       62            False
8    Albert       62            False
7     Marie       63            False
6    Thomas       64            False
4       Ada       92             True
3     Billy       96             True
10    Isaac       99             True
1      Jane       79             True
5      John       76             True
9   Richard       80             True
0     Sally       85             True
2     Suzie       74             True

 * d. Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [29]:
df[['name', 'english', 'passing_english']].sort_values(['passing_english', 'english'])

name  english  passing_english
8    Albert       62            False
11     Alan       62            False
7     Marie       63            False
6    Thomas       64            False
2     Suzie       74             True
5      John       76             True
1      Jane       79             True
9   Richard       80             True
0     Sally       85             True
4       Ada       92             True
3     Billy       96             True
10    Isaac       99             True

 * e. Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [34]:
df['overall_grade'] = (df['english'] + df['math'] +df['reading']) / 3

In [35]:
df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True      75.666667
1      Jane    88       79       67             True      78.000000
2     Suzie    94       74       95             True      87.666667
3     Billy    98       96       88             True      94.000000
4       Ada    77       92       98             True      89.000000
5      John    79       76       93             True      82.666667
6    Thomas    82       64       81            False      75.666667
7     Marie    93       63       90            False      82.000000
8    Albert    92       62       87            False      80.333333
9   Richard    69       80       94             True      81.000000
10    Isaac    92       99       93             True      94.666667
11     Alan    92       62       72            False      75.333333

2. Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [36]:
mpg.info

<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

* How many rows and columns are there?

234 rows x 11 columns

* What are the data types of each column?

In [42]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


* Summarize the dataframe with .info and .describe

In [44]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

In [ ]:
# interesting.... mpg.info and mpg.decribe give the same output, while mpg.info() and mpg.describe() are different

* Rename the cty column to city.

In [45]:
mpg.rename(columns={'cty': 'city'} , inplace=True)
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

* Rename the hwy column to highway.

In [48]:
mpg.rename(columns={'hwy': 'highway'}, inplace=True)
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

* Do any cars have better city mileage than highway mileage?

In [49]:
mpg[mpg['city'] > mpg['highway']]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

NOPE

* Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

In [50]:
mpg['mileage_difference'] = mpg['highway'] - mpg['city']
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

* Which car (or cars) has the highest mileage difference?

In [53]:
mpg.sort_values('mileage_difference', ascending=False).head()

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   
1           audi          a4    1.8  1999    4  auto(l5)   f    18       29   
229   volkswagen      passat    1.8  1999    4  auto(l5)   f    18       29   
36     chevrolet      malibu    3.5  2008    6  auto(l4)   f    18       29   

    fl       class  mileage_difference  
107  c  subcompact                  12  
223  d  subcompact                  12  
1    p     compact                  11  
229  p     midsize                  11  
36   r     midsize                  11

* Which compact class car has the lowest highway mileage? The best?

In [65]:
mpg[mpg['class'] == 'compact'].sort_values('highway')[:1]

manufacturer  model  displ  year  cyl     trans drv  city  highway fl  \
220   volkswagen  jetta    2.8  1999    6  auto(l4)   f    16       23  r   

       class  mileage_difference  
220  compact                   7

In [67]:
mpg[mpg['class'] == 'compact'].sort_values('highway')[-1:]

manufacturer  model  displ  year  cyl       trans drv  city  highway fl  \
213   volkswagen  jetta    1.9  1999    4  manual(m5)   f    33       44  d   

       class  mileage_difference  
213  compact                  11

* Create a column named average_mileage that is the mean of the city and highway mileage.

In [70]:
mpg['average_mileage'] = (mpg['city'] + mpg['highway'])/2

In [71]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

* Which dodge car has the best average mileage? The worst?

In [75]:
mpg[mpg['manufacturer']=='dodge'].sort_values('average_mileage', ascending=False)[:1]

manufacturer        model  displ  year  cyl     trans drv  city  highway  \
38        dodge  caravan 2wd    2.4  1999    4  auto(l3)   f    18       24   

   fl    class  mileage_difference  average_mileage  
38  r  minivan                   6             21.0

In [78]:
mpg[mpg['manufacturer']=='dodge'].sort_values('average_mileage', ascending=False)[-1:]

manufacturer                model  displ  year  cyl       trans drv  city  \
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   

    highway fl   class  mileage_difference  average_mileage  
70       12  e  pickup                   3             10.5

3. Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

In [81]:
mammals = data('Mammals')

In [82]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [83]:
mammals.describe()

weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

In [84]:
mammals.head()

weight  speed  hoppers  specials
1  6000.0   35.0    False     False
2  4000.0   26.0    False     False
3  3000.0   25.0    False     False
4  1400.0   45.0    False     False
5   400.0   70.0    False     False

In [85]:
mammals.info

<bound method DataFrame.info of        weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    13

* How many rows and columns are there?

107 x 4

* What are the data types?

In [88]:
mammals.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

* Summarize the dataframe with .info and .describe

In [89]:
mammals.info

<bound method DataFrame.info of        weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    13

In [90]:
mammals.describe

<bound method NDFrame.describe of        weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    

In [91]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [92]:
mammals.describe()

weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

* What is the the weight of the fastest animal?

In [93]:
mammals['weight'].max()

6000.0

* What is the overal percentage of specials?

In [104]:
10000 / mammals['weight'].count() / mammals['weight'][mammals['specials']==True].count()

9.345794392523365

* How many animals are hoppers that are above the median speed? What percentage is this?

In [107]:
mammals['speed'].median()

48.0

In [106]:
mammals[mammals['hoppers']==True]

weight  speed  hoppers  specials
82    0.056   21.0     True     False
85    0.035   32.0     True     False
86    0.035   14.0     True     False
96    4.600   64.0     True     False
97    4.400   72.0     True     False
98    4.000   72.0     True     False
99    3.500   56.0     True     False
100   2.000   64.0     True     False
101   1.900   56.0     True     False
102   1.500   50.0     True     False
103   1.500   40.0     True     False

In [114]:
mammals[mammals['hoppers']==True][mammals['speed']>mammals['speed'].median()]

/var/folders/n7/p9r5pxl53g99g0h4q1_h50fr0000gn/T/ipykernel_79267/1644194577.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  mammals[mammals['hoppers']==True][mammals['speed']>mammals['speed'].median()]


weight  speed  hoppers  specials
96      4.6   64.0     True     False
97      4.4   72.0     True     False
98      4.0   72.0     True     False
99      3.5   56.0     True     False
100     2.0   64.0     True     False
101     1.9   56.0     True     False
102     1.5   50.0     True     False

### Awesome Bonus
For much more practice with pandas, Go to https://github.com/guipsamora/pandas_exercises and clone the repo down to your laptop. To clone a repository: - Copy the SSH address of the repository - cd ~/codeup-data-science - Then type git clone git@github.com:guipsamora/pandas_exercises.git - Now do cd pandas_exercises on your terminal. - Type git remote remove origin, so you won't accidentally try to push your work to guipsamora's repo.

Congratulations! You have cloned guipsamora's pandas exercises to your computer. Now you need to make a new, blank, repository on GitHub.

Go to https://github.com/new to make a new repo. Name it pandas_exercises.
DO NOT check any check boxes. We need a blank, empty repo.
Finally, follow the directions to "push an existing repository from the command line" so that you can push up your changes to your own account.
Now do your own work, add it, commit it, and push it!